### PR TITLE
Introduce OnCompleted to the transport seam

### PR DESCRIPTION
**Pull request by andreasohlund**
This allows https://github.com/Particular/NServiceBus.Metrics/issues/174 to be fixed.

As a side effect, we can also improve our transport tests that now can be made less brittle and faster (separate PR will be raised)

### CODE REVIEW

File: src/NServiceBus.AcceptanceTests/Routing/MessageDrivenSubscriptions/When_missing_publisher_information.cs [hunk 1]
```diff
@@ -8,7 +8,7 @@
     using Logging;
     using NUnit.Framework;
 
-    public class When_autosubscribe_with_missing_publisher_information : NServiceBusAcceptanceTest
+    public class When_missing_publisher_information : NServiceBusAcceptanceTest
```
Comment (andreasohlund) [inline src/NServiceBus.AcceptanceTests/Routing/MessageDrivenSubscriptions/When_missing_publisher_information.cs:11]: Couldn't build locally since this generated a to long path

**Comment by andreasohlund** [comment]
@DavidBoike @adamralph should I refactor the transport tests as part of this PR?

**Comment by DavidBoike** [comment]
👍

File: src/NServiceBus.Core/Transports/Learning/LearningTransportMessagePump.cs [hunk 2]
```diff
@@ -260,11 +266,20 @@ await ProcessFile(transaction, messageId, messageProcessingCancellationToken)
                     log.Debug($"Failure while trying to complete receive transaction for  {filePath}({transaction.FileToProcess})" + filePath, ex);
                 }
 
+                try
+                {
+                    await onCompleted(new CompleteContext(messageId, wasAcknowledged, headers ?? new Dictionary<string, string>(), startedAt, DateTimeOffset.UtcNow, processingContext), messageProcessingCancellationToken).ConfigureAwait(false);
+                }
+                catch (Exception ex)
+                {
+                    log.Debug($"Failure when invoking OnComplete for {filePath}({transaction.FileToProcess})" + filePath, ex);
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Transports/Learning/LearningTransportMessagePump.cs:275]: Just continued to do what all other code does in this method but DEBUG feels off to me? perhaps WARN?

File: src/NServiceBus.Core/Transports/CompleteContext.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+        /// <summary>
+        /// The time processing completed.
+        /// </summary>
+        public DateTimeOffset CompletedAt { get; }
+
+        /// <summary>
+        /// A <see cref="ContextBag" /> which can be used to extend the current object.
+        /// </summary>
+        public ContextBag Extensions { get; }
```
Comment (adamralph) [inline src/NServiceBus.Core/Transports/CompleteContext.cs:64]: Why is this needed? FWIW, I tried removing it and everything still worked.

File: src/NServiceBus.Core/Receiving/ProcessingCompleted.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+            WasAcknowledged = wasAcknowledged;
+            Headers = headers;
+            StartedAt = startedAt;
+            CompletedAt = completedAt;
+        }
+
+        /// <summary>
+        /// The ID of the message.
+        /// </summary>
+        public string MessageId;
```
Comment (adamralph) [inline src/NServiceBus.Core/Receiving/ProcessingCompleted.cs:26]:
```suggestion
        public string MessageId { get; }
```

File: src/NServiceBus.Core/Receiving/ProcessingCompleted.cs [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+        /// <summary>
+        /// True if the message was acknowledged and removed from the queue.
+        /// </summary>
+        public bool WasAcknowledged;
```
Comment (adamralph) [inline src/NServiceBus.Core/Receiving/ProcessingCompleted.cs:31]:
```suggestion
        public bool WasAcknowledged { get; }
```

File: src/NServiceBus.Core/Receiving/ProcessingCompleted.cs (hunk 4 above)
Comment (adamralph) [inline src/NServiceBus.Core/Receiving/ProcessingCompleted.cs:26]: I see that `IncomingMessage`, which is a property on `ReceivePipelineCompleted`, has both `MessageId` and `NativeMessageId`. Which one is this?

File: src/NServiceBus.Core/Receiving/ProcessingCompleted.cs [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+        /// <summary>
+        /// The message headers.
+        /// </summary>
+        public Dictionary<string, string> Headers { get; }
```
Comment (adamralph) [inline src/NServiceBus.Core/Receiving/ProcessingCompleted.cs:36]: Why is this required?

File: src/NServiceBus.Core/Receiving/ProcessingCompleted.cs (hunk 5 above)
Comment (adamralph) [inline src/NServiceBus.Core/Receiving/ProcessingCompleted.cs:31]: I guess this is required because the event is being raised regardless of whether the message was successfully processed or not, whereas `ReceivePipelineCompleted` is only raised if the receive pipeline does not throw an exception?

File: src/NServiceBus.Core/Receiving/ReceiveComponent.cs [hunk 7]
```diff
@@ -174,7 +175,7 @@ partial class ReceiveComponent
                     var satelliteRecoverabilityExecutor = recoverabilityExecutorFactory.Create(satellite.RecoverabilityPolicy, satellite.ReceiveAddress);
 
                     await satellitePump.Initialize(satellite.RuntimeSettings, satellitePipeline.Invoke,
-                        satelliteRecoverabilityExecutor.Invoke, cancellationToken).ConfigureAwait(false);
+                        satelliteRecoverabilityExecutor.Invoke, (_, __) => Task.CompletedTask, cancellationToken).ConfigureAwait(false);
```
Comment (adamralph) [inline src/NServiceBus.Core/Receiving/ReceiveComponent.cs:178]: Are we sure this is OK?

File: src/NServiceBus.Core/Transports/Learning/LearningTransportMessagePump.cs (hunk 2 above)
Comment (adamralph) [reply src/NServiceBus.Core/Transports/Learning/LearningTransportMessagePump.cs:275]: I wonder if this shouldn't be handled in Core instead?

File: src/NServiceBus.Core/Receiving/MainOnCompleteExecutor.cs [hunk 8]
```diff
@@ -0,0 +1,21 @@
+﻿namespace NServiceBus
+{
+    using System.Threading;
+    using System.Threading.Tasks;
+    using NServiceBus.Transport;
+
+    class MainOnCompleteExecutor
```
[the remark this answers is not in the document]
Comment (andreasohlund) [reply src/NServiceBus.Core/Receiving/MainOnCompleteExecutor.cs:7]: I added "Main" to follow the convention of https://github.com/Particular/NServiceBus/blob/master/src/NServiceBus.Core/Pipeline/MainPipelineExecutor.cs since this only triggers on the main receiver and not on the satellite receivers?

File: src/NServiceBus.Core/Transports/Learning/LearningTransportMessagePump.cs (hunk 2 above)
Comment (andreasohlund) [reply src/NServiceBus.Core/Transports/Learning/LearningTransportMessagePump.cs:275]: Not sure, we do try catch the calls to both OnError and OnMessage?

File: src/NServiceBus.Core/Receiving/ProcessingCompleted.cs [hunk 9]
```diff
@@ -0,0 +1,48 @@
+﻿namespace NServiceBus
+{
+    using System;
+    using System.Collections.Generic;
+
+    /// <summary>
+    /// Event raised when the transport has completed processing a message.
+    /// </summary>
+    public class ProcessingCompleted
```
[the remark this answers is not in the document]
Comment (andreasohlund) [reply src/NServiceBus.Core/Receiving/ProcessingCompleted.cs:9]: Yea, I like the name better. Will change

File: src/NServiceBus.Core/Receiving/ReceiveConfigExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+    /// <summary>
+    /// Provides access to processing complete notifications.
+    /// </summary>
+    public static class ReceiveConfigExtensions
+    {
+        /// <summary>
+        /// Subscribes to notifications for processing completed events.
+        /// </summary>
+        public static void OnMessageProcessingCompleted(this FeatureConfigurationContext featureConfigurationContext, Func<ProcessingCompleted, CancellationToken, Task> subscription)
```
[the remark this answers is not in the document]
Comment (andreasohlund) [reply src/NServiceBus.Core/Receiving/ReceiveConfigExtensions.cs:16]: The equivalent is here

https://github.com/Particular/NServiceBus/blob/master/src/NServiceBus.Core/Pipeline/Incoming/ReceivePipelineConfigExtensions.cs

it exposed it on the pipeline settings which are exposed both the features and end users. I figured to only expose to features for now to make the surface area minimum?

File: src/NServiceBus.Core/Receiving/ProcessingCompleted.cs (hunk 4 above)
Comment (andreasohlund) [reply src/NServiceBus.Core/Receiving/ProcessingCompleted.cs:26]: This is the native message id. I will rename

File: src/NServiceBus.Core/Receiving/ProcessingCompleted.cs (hunk 6 above)
Comment (andreasohlund) [reply src/NServiceBus.Core/Receiving/ProcessingCompleted.cs:36]: Since metrics and other users of this will need to access things in the headers. Metrics needs the sent time header and the message type header

File: src/NServiceBus.Core/Receiving/ProcessingCompleted.cs (hunk 5 above)
Comment (andreasohlund) [reply src/NServiceBus.Core/Receiving/ProcessingCompleted.cs:31]: Correct

File: src/NServiceBus.Core/Receiving/ReceiveComponent.cs (hunk 7 above)
Comment (andreasohlund) [reply src/NServiceBus.Core/Receiving/ReceiveComponent.cs:178]: Yes, we only trigger these events for the "main and instance receivers" not for the satellites

File: src/NServiceBus.Core/Transports/CompleteContext.cs (hunk 3 above)
Comment (andreasohlund) [reply src/NServiceBus.Core/Transports/CompleteContext.cs:64]: This is to provide a way for users to float context from OnMessage over to OnError and OnComplete but since onError doesn't seem to have it I will remove for now. Lets discuss this on the next sync

File: src/NServiceBus.Core/Receiving/ProcessingCompleted.cs (hunk 4 above)
Comment (andreasohlund) [reply src/NServiceBus.Core/Receiving/ProcessingCompleted.cs]: Renamed

File: src/NServiceBus.Core/Transports/CompleteContext.cs (hunk 3 above)
Comment (andreasohlund) [reply src/NServiceBus.Core/Transports/CompleteContext.cs]: Removed

File: src/NServiceBus.Core/Transports/CompleteContext.cs [hunk 11]
```diff
@@ -0,0 +1,66 @@
+﻿namespace NServiceBus.Transport
+{
+    using System;
+    using System.Collections.Generic;
+    using NServiceBus.Extensibility;
+
+    /// <summary>
+    /// Allows the transport to pass signal that a message has been completed.
+    /// </summary>
+    public class CompleteContext : IExtendable
```
[the remark this answers is not in the document]
Comment (andreasohlund) [reply src/NServiceBus.Core/Transports/CompleteContext.cs]: We need to talk this through but I have removed for now

File: src/NServiceBus.Core/Receiving/ProcessingCompleted.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (andreasohlund) [reply src/NServiceBus.Core/Receiving/ProcessingCompleted.cs]: Changed to `ReceiveCompleted`

**Comment by andreasohlund** [comment]
@adamralph this one is ready for another review

**Comment by andreasohlund** [comment]
Not sure if we should fix this here but its a bit off that the error context use `IncomingMessage` and the other context use id, headers, body as primitives

https://github.com/Particular/NServiceBus/pull/5936/files?file-filters%5B%5D=.cs&file-filters%5B%5D=.txt&hide-deleted-files=true#diff-cf366060edf38462d421b67f8c111af521772d1f0ef13b06f3dfd91e67f5744bL57

File: src/NServiceBus.Core/Receiving/ReceiveCompleted.cs [hunk 12]
```diff
@@ -0,0 +1,49 @@
+﻿namespace NServiceBus
+{
+    using System;
+    using System.Collections.Generic;
+
+    /// <summary>
+    /// Event raised when the transport has completed processing a message.
+    /// </summary>
+    public class ReceiveCompleted
+
```
Comment (DavidBoike) [inline src/NServiceBus.Core/Receiving/ReceiveCompleted.cs:10]:
```suggestion
```

File: src/NServiceBus.Core/Receiving/ReceiveCompleted.cs [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+            WasAcknowledged = wasAcknowledged;
+            Headers = headers;
+            StartedAt = startedAt;
+            CompletedAt = completedAt;
+        }
+
+        /// <summary>
+        /// The native message ID.
+        /// </summary>
+        public string NativeMessageId;
```
Comment (DavidBoike) [inline src/NServiceBus.Core/Receiving/ReceiveCompleted.cs:27]: Are NativeMessageId and WasAcknowledged public _fields_ for a reason?

**Review by DavidBoike — APPROVED** [review]

File: src/NServiceBus.Core/Receiving/ReceiveCompleted.cs (hunk 13 above)
Comment (andreasohlund) [reply src/NServiceBus.Core/Receiving/ReceiveCompleted.cs:26]: No, will make them props

File: src/NServiceBus.Core/Transports/Learning/LearningTransportMessagePump.cs (hunk 2 above)
Comment (adamralph) [reply src/NServiceBus.Core/Transports/Learning/LearningTransportMessagePump.cs:275]: I guess it should be logged as a warning, and if we're happy with adding this to the list of Things That Every Transport Must Do then let's just leave it here.

Incidentally, it seems off that we are swallowing the exception when the transaction fails to complete and just logging a debug message, but that's outside the scope of this PR.

File: src/NServiceBus.Core/Transports/ErrorContext.cs [hunk 14]
```diff
@@ -59,6 +67,6 @@ public ErrorContext(Exception exception, Dictionary<string, string> headers, str
         /// <summary>
         /// A collection of additional information provided by the transport.
         /// </summary>
-        public ReadOnlyContextBag Extensions { get; }
+        public ContextBag Extensions { get; }
```
Comment (adamralph) [inline src/NServiceBus.Core/Transports/ErrorContext.cs:70]: Given that the context needs to be propagated from message handling to error handling [for satellite receivers](https://github.com/Particular/NServiceBus/pull/5937), and [satellite receivers don't receive the receive completed event](https://github.com/Particular/NServiceBus/pull/5936/files#r579799291), I don't understand why they need to be able to write information into the context during error handling.

File: src/NServiceBus.Core/Transports/MessageContext.cs [hunk 15]
```diff
@@ -11,30 +11,30 @@ public partial class MessageContext : IExtendable
...
+            TransportMessageId = transportMessageId;
             Extensions = context;
             TransportTransaction = transportTransaction;
         }
 
         /// <summary>
         /// The native id of the message.
         /// </summary>
-        public string MessageId { get; }
+        public string TransportMessageId { get; }
```
Comment (adamralph) [inline src/NServiceBus.Core/Transports/MessageContext.cs:37]: Taking stock of the various ID's:

```c#
MessageContext.TransportMessageId

ErrorContext.Message.MessageId
ErrorContext.Message.NativeMessageId

CompleteContext.TransportMessageId
```

Something seems off with this. Does any ID flow through all three contexts? If so, I would expect it to be named the same in each.

File: src/NServiceBus.Core/Transports/ErrorContext.cs [hunk 16]
```diff
@@ -18,9 +18,17 @@ public class ErrorContext
...
+        public ErrorContext(Exception exception, Dictionary<string, string> headers, string transportMessageId, byte[] body, TransportTransaction transportTransaction, int immediateProcessingFailures, ContextBag context)
         {
+            Guard.AgainstNull(nameof(exception), exception);
+            Guard.AgainstNull(nameof(headers), headers);
+            Guard.AgainstNullAndEmpty(nameof(transportMessageId), transportMessageId);
+            Guard.AgainstNull(nameof(body), body);
+            Guard.AgainstNull(nameof(transportTransaction), transportTransaction);
+            Guard.AgainstNegative(nameof(immediateProcessingFailures), immediateProcessingFailures);
+            Guard.AgainstNull(nameof(context), context);
+
```
Comment (adamralph) [inline src/NServiceBus.Core/Transports/ErrorContext.cs:31]: I guess these guards are unrelated to the purpose of this PR and just Stuff That Should Always Have Been Here?

File: src/NServiceBus.Core/Transports/ErrorContext.cs [hunk 17]
```diff
@@ -18,9 +18,17 @@ public class ErrorContext
...
         /// <param name="immediateProcessingFailures">Number of failed immediate processing attempts.</param>
-        /// <param name="context">A <see cref="ReadOnlyContextBag" /> which can be used to extend the current object.</param>
-        public ErrorContext(Exception exception, Dictionary<string, string> headers, string transportMessageId, byte[] body, TransportTransaction transportTransaction, int immediateProcessingFailures, ReadOnlyContextBag context)
+        /// <param name="context">A <see cref="ContextBag" /> which can be used to extend the current object.</param>
+        public ErrorContext(Exception exception, Dictionary<string, string> headers, string transportMessageId, byte[] body, TransportTransaction transportTransaction, int immediateProcessingFailures, ContextBag context)
         {
+            Guard.AgainstNull(nameof(exception), exception);
+            Guard.AgainstNull(nameof(headers), headers);
+            Guard.AgainstNullAndEmpty(nameof(transportMessageId), transportMessageId);
+            Guard.AgainstNull(nameof(body), body);
```
Comment (adamralph) [inline src/NServiceBus.Core/Transports/ErrorContext.cs:27]: I don't think we should be checking these here. All we are doing is passing them to `IncomingMessage`. Indeed, looking inside the constructor, we can see that these checks here are redundant.
```suggestion
```

**Review by adamralph — COMMENTED** [review]
> Not sure if we should fix this here but its a bit off that the error context use IncomingMessage and the other context use id, headers, body as primitives
> 
> https://github.com/Particular/NServiceBus/pull/5936/files?file-filters%5B%5D=.cs&file-filters%5B%5D=.txt&hide-deleted-files=true#diff-cf366060edf38462d421b67f8c111af521772d1f0ef13b06f3dfd91e67f5744bL57

I agree that aligning `ErrorContext` with existing contexts is outside the scope of this PR. For the new context we are introducing, we should ensure that we are using the "correct" approach.

File: src/NServiceBus.Core/Receiving/ReceiveConfigExtensions.cs [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+
+    /// <summary>
+    /// Provides access to processing complete notifications.
+    /// </summary>
+    public static class ReceiveConfigExtensions
+    {
+        /// <summary>
+        /// Subscribes to notifications for receive completed events.
+        /// </summary>
+        public static void OnMessageReceiveCompleted(this FeatureConfigurationContext featureConfigurationContext, Func<ReceiveCompleted, CancellationToken, Task> subscription)
```
[the remark this answers is not in the document]
Comment (andreasohlund) [reply src/NServiceBus.Core/Receiving/ReceiveConfigExtensions.cs:16]: Yes, metrics needs this see https://github.com/Particular/NServiceBus.Metrics/blob/master/src/NServiceBus.Metrics/ProbeBuilders/CriticalTimeProbeBuilder.cs#L17

File: src/NServiceBus.Core/Transports/ErrorContext.cs (hunk 16 above)
Comment (andreasohlund) [reply src/NServiceBus.Core/Transports/ErrorContext.cs:31]: Correct

File: src/NServiceBus.Core/Transports/MessageContext.cs (hunk 15 above)
Comment (andreasohlund) [reply src/NServiceBus.Core/Transports/MessageContext.cs:37]: Will rename to NativeMessageId

File: src/NServiceBus.Core/Transports/ErrorContext.cs (hunk 14 above)
Comment (andreasohlund) [reply src/NServiceBus.Core/Transports/ErrorContext.cs:67]: >satellite receivers don't receive the receive completed event

That is just because we haven't exposed it yet. Since context can flow from OnMessage to OnError I'd say its expected for it to be able to flow to OnComplete as well

File: src/NServiceBus.Core/Receiving/ReceiveCompleted.cs [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+
+        /// <summary>
+        /// The native message ID.
+        /// </summary>
+        public string NativeMessageId { get; }
+
+        /// <summary>
+        /// True if the message was acknowledged and removed from the queue.
+        /// </summary>
+        public bool WasAcknowledged { get; }
```
[the remark this answers is not in the document]
Comment (andreasohlund) [reply src/NServiceBus.Core/Receiving/ReceiveCompleted.cs:31]: Yes metrics (and our tests) need to know if the message was acked or not. In the metics case the "MessageProcessed" metrics should not be emitted if the message was rolled back.

>Having said that, we could introduce all that later, and just deprecate this property. Or it could remain as a convenience property which returns true when the outcome was "message was processed" or "message was moved to error queue".

Agree, this can be added later

File: src/NServiceBus.Core/Transports/CompleteContext.cs [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+
+        /// <summary>
+        /// The time processing completed.
+        /// </summary>
+        public DateTimeOffset CompletedAt { get; }
+
+        /// <summary>
+        /// A collection of additional information for this receive operation.
+        /// </summary>
+        public ReadOnlyContextBag Extensions { get; }
```
[the remark this answers is not in the document]
Comment (andreasohlund) [reply src/NServiceBus.Core/Transports/CompleteContext.cs:65]: Its needed for things to be consistent. We support passing context to OnError so it should make sense to pass it to OnComplete as well see https://github.com/Particular/NServiceBus/pull/5936/files#diff-78ac3141bdcecc259c5301c559bc879e8c77d850bff241e9ea8e0d9136de5a4dR14

File: src/NServiceBus.Core/Receiving/ReceiveConfigExtensions.cs (hunk 18 above)
[the remark this answers is not in the document]
Comment (andreasohlund) [reply src/NServiceBus.Core/Receiving/ReceiveConfigExtensions.cs:16]: Yea, will rename

File: src/NServiceBus.Core/Receiving/ReceiveComponent.Settings.cs [hunk 21]
```diff
@@ -63,6 +63,8 @@ public PushRuntimeSettings PushRuntimeSettings
 
             public Notification<ReceivePipelineCompleted> PipelineCompletedSubscribers => settings.GetOrCreate<Notification<ReceivePipelineCompleted>>();
 
+            public Notification<ReceiveCompleted> ProcessingCompletedSubscribers => settings.GetOrCreate<Notification<ReceiveCompleted>>();
```
[the remark this answers is not in the document]
Comment (andreasohlund) [reply src/NServiceBus.Core/Receiving/ReceiveComponent.Settings.cs:66]: 👍

File: src/NServiceBus.Core/Receiving/ReceiveComponent.Configuration.cs [hunk 22]
```diff
@@ -112,6 +115,8 @@ public void AddSatelliteReceiver(string name, string transportAddress, PushRunti
 
             public Notification<ReceivePipelineCompleted> PipelineCompletedSubscribers;
 
+            public Notification<ReceiveCompleted> ProcessingCompletedSubscribers;
```
[the remark this answers is not in the document]
Comment (andreasohlund) [reply src/NServiceBus.Core/Receiving/ReceiveComponent.Configuration.cs:118]: 👍

File: src/NServiceBus.Core/Receiving/MainOnCompleteExecutor.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (andreasohlund) [reply src/NServiceBus.Core/Receiving/MainOnCompleteExecutor.cs:7]: Lets discuss tomorrow

File: src/NServiceBus.Core/Receiving/ReceiveComponent.cs [hunk 23]
```diff
@@ -147,19 +147,20 @@ partial class ReceiveComponent
...
+                recoverability.Invoke, onCompleteExecutor.Invoke, cancellationToken).ConfigureAwait(false);
             receivers.Add(mainPump);
 
             if (transportInfrastructure.Receivers.TryGetValue(InstanceSpecificReceiverId, out var instanceSpecificPump))
             {
                 var instanceSpecificRecoverabilityExecutor = recoverabilityExecutorFactory.CreateDefault(configuration.InstanceSpecificQueue);
 
                 await instanceSpecificPump.Initialize(configuration.PushRuntimeSettings, mainPipelineExecutor.Invoke,
-                    instanceSpecificRecoverabilityExecutor.Invoke, cancellationToken).ConfigureAwait(false);
+                    instanceSpecificRecoverabilityExecutor.Invoke, onCompleteExecutor.Invoke, cancellationToken).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (andreasohlund) [reply src/NServiceBus.Core/Receiving/ReceiveComponent.cs]: fixed

File: src/NServiceBus.Core/Transports/Learning/LearningTransportMessagePump.cs (hunk 2 above)
Comment (andreasohlund) [reply src/NServiceBus.Core/Transports/Learning/LearningTransportMessagePump.cs]: now logging WARN

**Comment by andreasohlund** [comment]
@adamralph addressed the changes we talked about

@adamralph  +  @DavidBoike introduced the new flag that we talked about to indicate that the message processing failed. Not super happy with both the name and implementation so feel free to refactor and/or rename, see https://github.com/Particular/NServiceBus/pull/5936/commits/914d54bdc637bf85a679e2453fc6f0163eb91f0a

**Comment by DavidBoike** [comment]
Looked through the last bit. Removed a duplicate "the". Let's merge this!

File: src/NServiceBus.Core/Receiving/ReceiveComponent.Configuration.cs (hunk 22 above)
[the remark this answers is not in the document]
Comment (adamralph) [reply src/NServiceBus.Core/Receiving/ReceiveComponent.Configuration.cs:118]: resolved in https://github.com/Particular/NServiceBus/pull/5945
